plugins/hls-hlint-plugin/src/Ide/Plugin/Hlint.hs

```diff
@@ -268,20 +268,21 @@ getHlintSettingsRule usage =
 -- ---------------------------------------------------------------------
 
 codeActionProvider :: PluginMethodHandler IdeState TextDocumentCodeAction
-codeActionProvider ideState plId (CodeActionParams _ _ docId _ context) = Right . LSP.List . map InR <$> liftIO getCodeActions
+codeActionProvider ideState plId (CodeActionParams _ _ docId caRange context) = Right . LSP.List . map InR <$> liftIO getCodeActions
   where
 
     getCodeActions = do
         diags <- getDiagnostics ideState
         let docNfp = toNormalizedFilePath' <$> uriToFilePath' (docId ^. LSP.uri)
-            numHintsInDoc = length
-              [d | (nfp, _, d) <- diags
+            numHintsInRange = length
+              [d | (nfp, _, d@LSP.Diagnostic{_range=range}) <- diags
                  , validCommand d
                  , Just nfp == docNfp
+                 , subRange range caRange
               ]
         -- We only want to show the applyAll code action if there is more than 1
         -- hint in the current document
-        if numHintsInDoc > 1 then do
+        if numHintsInRange > 1 then do
           pure $ applyAllAction:applyOneActions
```

Comment:
I think you should make sure that `applyAllAction` is *not* always prepended to `applyOneActions`, rather than changing `numHintsInDoc`.

Reply:
Isn't it okay since it's only used to determine if we should prepend to `applyOneActions`? And since I add more filtering, `numHintsInRange <= numHintsInDoc` is always true so the condition in the change already includes `numHintsInDoc > 1` too.

Reply:
Okay, I just I realized that `Apply all hints` does not mean applying all hints for those range, but it means applying all the hints in doc. I've changed the condition

Reply:
We should show `applyAllAction` iff:

* there are more than one hlint diagnostics in the document
* there is at least one hlint diagnostic in the current code action context

right? So based on `numHintsInDoc`, which has already implemented the logic for the first condition, you just need take a look at diagnostics in the current code action context.

Reply:
Right, I tried just adding `not (null diags)` and it works. The problem now is that lsp-test sends all the diagnostics instead of only those within [range]( https://hackage.haskell.org/package/lsp-test-0.13.0.0/docs/src/Language.LSP.Test.html#getCodeActionContext).

Reply:
`not (null diags)` is not good enough, since diagnostics from sources other than hlint may trigger the show of `applyAllAction` -- you could use `validCommand` to filter out code actions we don't care about. For test, sorry I don't understant clearly what you said. Can't `getCodeActions` work?

Reply:
> Can't getCodeActions work?

Yes, in the test file if I call `getCodeActions` for the last line, it will actually send the whole diagnostic as the context to the server, not only those that overlaps with the range.

Reply:
I updated the code to check if there's hlint diagnostics within the context passed. Test will fail, but this is merged https://github.com/bubba/lsp-test/pull/91. Maybe we can wait for new lsp-test version?

Reply:
I don't think we can necessarily rely on clients filling in the diagnostics in the context. Here's what the spec says:
```
         /**
	 * An array of diagnostics known on the client side overlapping the range
	 * provided to the `textDocument/codeAction` request. They are provided so
	 * that the server knows which errors are currently presented to the user
	 * for the given range. There is no guarantee that these accurately reflect
	 * the error state of the resource. The primary parameter
	 * to compute code actions is the provided range.
	 */
```

So it seems to suggest that you should primarily rely on the provided range.

Reply:
I might be wrong, but my understanding here is:

> They are provided so that the server knows which errors are currently presented to the user for the given range

The client gives us the diagnostics that is shown to the user. If in the meantime the diagnostics changes, shouldn't the client issue the request again? We don't know for sure if the new diagnostics changes is already shown to the user (or even received by the client) and we might suggest actions that mismatch with what the user sees.

Reply:
> There is no guarantee that these accurately reflect the error state of the resource. The primary parameter to compute code actions is the provided range.

IMO this is the key bit.

Reply:
Yeah, but I'm not sure if this is possible:
1. Client receives and shows diagnostic `dA`
2. Client issues code action request with `dA` as context
3. New diagnostic `dB` is sent from the server
4. Server sends code action based on `dB` (since it's latest state)
5. Client receives only code action suggestion but not the diagnostic message `dB` 

In this case client will show diagnostic `dA` with code action `dB`.

Reply:
Well, consider that AFAICT this is not mandatory, so if a client simply hasn't implemented adding the extra diagnostics, then this feature will not work, despite us having access to a perfectly good range we can use.

Reply:
I don't think they're optional? The definition doesn't have `?` in the name.